src/linux/routing/queueing/internal.hpp

```diff
@@ -118,7 +120,69 @@ Try<Netlink<struct rtnl_qdisc>> encodeDiscipline(
 }
 
 /////////////////////////////////////////////////
-// Helpers for internal APIs.
+// Helpers for {en}decoding queueing classes.
+/////////////////////////////////////////////////
+
+namespace cls {
+// Forward declaration. Each type of queueing class needs to implement
+// this function to encode its type specific configurations into the
+// libnl queueing class (rtnl_class).
+template <typename Config>
+Try<Nothing> encode(
+    const Netlink<struct rtnl_class>& cls,
+    const Config& config);
+
+
+// Forward declaration. Each type of queueing class needs to implement
+// this function to decode its type specific configurations from the
+// libnl queueing class (rtnl_class). Returns None if the libnl
+// queueing class does not match the specific queueing class type.
+template <typename Config>
+Result<Config> decode(const Netlink<struct rtnl_class>& cls);
+
+
+// Encodes a queueing class (in our representation) to a libnl
+// queueing class (rtnl_class). We use templating here so that it works
+// for any type of queueing class.
+template <typename Config>
+Try<Netlink<struct rtnl_class>> encode(
+    const Netlink<struct rtnl_link>& link,
+    const Class<Config>& config)
+{
+  struct rtnl_class* c = rtnl_class_alloc();
+  if (c == nullptr) {
+    return Error("Failed to allocate a libnl class");
+  }
+
+  Netlink<struct rtnl_class> cls(c);
+
+  rtnl_tc_set_link(TC_CAST(cls.get()), link.get());
+  rtnl_tc_set_parent(TC_CAST(cls.get()), config.parent.get());
+
+  if (config.handle.isSome()) {
+    rtnl_tc_set_handle(TC_CAST(cls.get()), config.handle.get().get());
+  }
+
+  int error = rtnl_tc_set_kind(TC_CAST(cls.get()), config.kind.c_str());
+  if (error != 0) {
+    return Error(
+        "Failed to set the kind of the queueing class: " +
+        std::string(nl_geterror(error)));
+  }
+
+  // Perform queueuing class specific encoding.
+  Try<Nothing> encoding = encode(cls, config.config);
+  if (encoding.isError()) {
+    return Error("Failed to encode the queueing class: " + encoding.error());
+  }
```

Comment:
along the same line of thinking, aren't these error cases leaking the memory..? we've done the rtnl_class_alloc() but we're erroring out here and no one is going to free it?